include/wil/toolhelp32.h

```diff
@@ -0,0 +1,133 @@
+#include "resource.h"
+#ifndef __WIL_TOOLHELP32_INCLUDED
+#define __WIL_TOOLHELP32_INCLUDED
+#include <TlHelp32.h>
+#include <processthreadsapi.h>
+namespace wil
+{
+namespace details
+{
+    template <typename TEntry, typename TEnumApi, typename TCallback>
+    void do_enum_snapshot(HANDLE handle, TEntry& entry, TEnumApi&& enumApiFirst, TEnumApi&& enumApiNext, TCallback&& callback)
+    {
+        using result_t = decltype(callback(TEntry{}));
+        bool enumResult = enumApiFirst(handle, &entry);
+        if (!enumResult)
+            return;
+
+        do
+        {
+            if constexpr (wistd::is_void_v<result_t>)
+            {
+                callback(entry);
+            }
+            else if constexpr (wistd::is_same_v<result_t, bool>)
+            {
+                if (callback(entry))
+                    return;
+            }
+            else
+            {
+                static_assert(
+                    [] {
+                        return false;
+                    }(),
+                    "Callback must return void or bool");
+            }
+            enumResult = enumApiNext(handle, &entry);
+        } while (enumResult);
+    }
+} // namespace details
+
+template <typename TCallback>
+void for_each_process(TCallback&& callback)
+{
+    PROCESSENTRY32 entry{};
+    entry.dwSize = sizeof(entry);
+    details::do_enum_snapshot(
+        unique_handle{CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0)}.get(),

```

Comment:
All of these functions - `CreateToolhelp32Snapshot`, `Process32First`, etc. - can fail. Ideally the API would preserve & report failures to the caller

Reply:
@dunhor Agreed. Would early returning `false` in case of anything fails (something like this) be sufficient?
```cpp
template<...>
bool for_each_process(...)
{
    unique_handle h{Create...()};
    if (!h) return false;
    //...
}
```

Reply:
WIL is pretty heavily standardized on using `HRESULT` as the error type communicated back to callers. The general pattern is to have three functions: a base `_nothrow` function, a `_failfast` version that fails fast on error, and an exception throwing function. So in this case, you'd basically have the three functions:
```c++
HRESULT for_each_process_nothrow(...) { ... }
void for_each_process_failfast(...) { FAIL_FAST_IF_FAILED(for_each_process_nothrow(...); }
#if WIL_ENABLE_EXCEPTION
void for_each_process(...) { THROW_IF_FAILED(for_each_process_nothrow(...); }
#endif
```
Given that your function accepts an arbitrary lambda from the user that might throw, `for_each_process_nothrow` probably shouldn't be made `noexcept`, making the function name a little misleading, but IMO that's probably fine.